### PR TITLE
AVX2 implementation for tanh

### PR DESCRIPTION
**Pull request by vedanuj**
This PR addresses #4984. Adding AVX2 implementation for `tanh` function

**Before 7a28292**
```
In [12]: x = torch.randn(10000, 10000)

In [13]: %time _ = torch.tanh(x)
CPU times: user 796 ms, sys: 126 ms, total: 923 ms
Wall time: 925 ms

In [15]: x = torch.randn(1000, 1000)

In [16]: %time _ = torch.tanh(x)
CPU times: user 7.27 ms, sys: 916 µs, total: 8.19 ms
Wall time: 7.17 ms
```

**After df69faf**
```
In [36]: x = torch.randn(10000, 10000)

In [37]: %time _ = torch.tanh(x)
CPU times: user 282 ms, sys: 153 ms, total: 434 ms
Wall time: 436 ms

In [51]: x = torch.randn(1000, 1000)

In [52]: %time _ = torch.tanh(x)
CPU times: user 2.11 ms, sys: 649 µs, total: 2.76 ms
Wall time: 1.95 ms

```

@zdevito @ezyang Please review 

------
Edit :: Modify perf numbers after moving to cmath `tanh` implementation.

### CODE REVIEW

File: aten/src/TH/vector/AVX2.cpp [hunk 1]
```diff
@@ -127,4 +127,30 @@ void THFloatVector_sigmoid_AVX2(float *y, const float *x, const ptrdiff_t n) {
...
+    YMM3 = _mm256_div_ps(one, YMM1);
+
+    _mm256_storeu_ps(
+        y + i,
+        _mm256_div_ps(_mm256_sub_ps(YMM0, YMM2), _mm256_add_ps(YMM0, YMM2)));
+    _mm256_storeu_ps(
+        y + i + 8,
+        _mm256_div_ps(_mm256_sub_ps(YMM1, YMM3), _mm256_add_ps(YMM1, YMM3)));
+  }
+  for (; i < (n); i++) {
```
Comment (goldsborough) [inline aten/src/TH/vector/AVX2.cpp]: Let's remove the spaces around `(n)`

File: aten/src/TH/vector/AVX2.cpp [hunk 2]
```diff
@@ -127,4 +127,30 @@ void THFloatVector_sigmoid_AVX2(float *y, const float *x, const ptrdiff_t n) {
   }
 }
 
+void THFloatVector_tanh_AVX2(float* y, const float* x, const ptrdiff_t n) {
+  ptrdiff_t i;
+  const __m256 one = _mm256_set1_ps(1.0f);
+  __m256 YMM0, YMM1, YMM2, YMM3;
```
Comment (goldsborough) [inline aten/src/TH/vector/AVX2.cpp]: Is there a reason to pre-declare the vector registers?

File: aten/src/TH/vector/AVX2.cpp [hunk 3]
```diff
@@ -127,4 +127,30 @@ void THFloatVector_sigmoid_AVX2(float *y, const float *x, const ptrdiff_t n) {
...
+  for (i = 0; i <= ((n)-16); i += 16) {
```
Comment (goldsborough) [inline aten/src/TH/vector/AVX2.cpp]: Can you add a comment for the formula this code is implementing?

**Review by goldsborough — COMMENTED** [review]
A few nits

File: aten/src/TH/vector/AVX2.cpp (hunk 1 above)
Comment (vedanuj) [reply aten/src/TH/vector/AVX2.cpp]: This is the formatting with `clang-format` we have in the root. I will remove the spaces.

File: aten/src/TH/vector/AVX2.cpp (hunk 2 above)
Comment (vedanuj) [reply aten/src/TH/vector/AVX2.cpp]: I was following what other functions in the `AVX2.cpp` was doing. We can declare them inside the loop as well.

**Review by apaszke — CHANGES_REQUESTED** [review]
We shouldn't use the simple formula to compute tanh, because `exp(x)` will easily blow up even for fairly small values of `x`. This PR will not only need comparisons in terms of perf, but also in terms of numerical accuracy.

Here's a script that shows that this method will quickly produce NaNs (and our current `.tanh()` doesn't suffer from this problem):
```python
import torch                                                                           
                                                                
def tanh(x):                                                    
    return (x.exp() - x.neg().exp()) / (x.exp() + x.neg().exp())
                                                                
x = torch.linspace(-200, 200, 8)                                
print(torch.stack([x, x.tanh(), tanh(x)]))                      
```

**Comment by vedanuj** [comment]
Adding `tanh256_ps` AVX2 implementation in `avx_mathfun.h` reproducing the formula in [cephes singles library](http://www.netlib.org/cephes/)  for 32-bit floating point tanh function.

Adding tests to check accuracy against numpy's `tanh`.

@apaszke @goldsborough  Please review

File: aten/src/TH/vector/avx_mathfun.h [hunk 4]
```diff
@@ -298,6 +299,89 @@ v8sf exp256_ps(v8sf x) {
   return y;
 }
 
+_PS256_CONST(tanh_0p625, 0.625f);
```
Comment (goldsborough) [inline aten/src/TH/vector/avx_mathfun.h]: I'm not so sure we want to put this here, `avx_mathfun.h` is like a library from http://software-lisc.fbk.eu/avx_mathfun/

Comment (vedanuj) [reply aten/src/TH/vector/avx_mathfun.h]: Since other functions in the file are also AVX2 versions of cephes library functions, I thought this will be a good place to add the `tanh256_ps` here. Please recommend which will be a good place to add this function.

**Comment by cpuhrsch** [comment]
@vedanuj  avx_mathfun.h has been [moved into ATen](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/native/cpu/avx_mathfun.h) and is also being used by [vec256](https://github.com/pytorch/pytorch/tree/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/cpu/vec256). It might be worthwhile implementing this operation in context of vec256 at this point rather than mathfun. Specifically [vec256_float.h](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/cpu/vec256/vec256_float.h) contains the specializations for floating point numbers. This is the place you'd want to move your implementation of tanh.

In order to get this to work, you'd then also need to add tanh to [vec256_double.h](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/cpu/vec256/vec256_double.h) and [vec256_base.h](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/cpu/vec256/vec256_base.h). This doesn't require a special vectorized implementation, the existing code is hopefully clear enough to see how to add this (just look at exp()). A first useful place to then make use of this is within [UnaryOps.cpp](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/native/UnaryOps.cpp). For that you'll only need to [add a tanh_kernel](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp) and place a macro in [UnaryOpsKernel.h](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/native/cpu/UnaryOpsKernel.h) and again [UnaryOps.cpp](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/native/UnaryOps.cpp) (you can look at the exp implementation again). Also don't forget to modify [Declarations.cwrap](https://github.com/pytorch/pytorch/blob/e4c0bb1809fd9bf9161392bfff7d06092adc224d/aten/src/ATen/Declarations.cwrap) by removing the reference to "tanh_" and changing "tanh" to "_tanh" and adding "cname: tanh" to the new "_tanh" (again you can look at "_exp"). 

UnaryOps will soon be moved entirely into ATen. For now it only is used for the contiguous cases, but support for non-contiguous tensors will follow [soon](https://github.com/pytorch/pytorch/pull/6119). At that point UnaryOps will only call into THC for CUDA support, but not anymore into TH. I'll then also port your implementation of tanh together with exp, log, etc. as part of the non-contiguous extension.

**Comment by vedanuj** [comment]
Moving the `tanh` AVX2 implementation to `vec256_float.h`. 

**Benchmark results with `tanh` vectorized::**

Before  b08dff8

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 31.751821756362915     type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 2.253028154373169      type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 2.5653436183929443     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 2.1067211627960205     type: torch.FloatTensor
```
After 1c7873b

```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 2.383469581604004      type: torch.FloatTensor
tanh:           size: 10^3      count: 1000     elapsed: 0.7881767749786377     type: torch.FloatTensor
tanh:           size: 10^4      count: 100      elapsed: 0.7657756805419922     type: torch.FloatTensor
tanh:           size: 10^5      count: 10       elapsed: 0.5897262096405029     type: torch.FloatTensor
```

@cpuhrsch @apaszke @goldsborough Please review.

File: test/test_torch.py [hunk 5]
```diff
@@ -299,6 +299,16 @@ def test_tan(self):
...
 
+        # test with high values
+        x = np.linspace(-2000000000000.0, 2000000000000.0, 100)
+        y = torch.FloatTensor(x)
+        self.assertEqual(np.tanh(y), torch.tanh(y))
+
+        # test with fractional values
+        x = np.linspace(-1.0, 1.0, 100)
+        y = torch.FloatTensor(x)
+        self.assertEqual(np.tanh(y), torch.tanh(y))
```
Comment (apaszke) [inline test/test_torch.py]: You need to make a separate test for the numpy part and guard it with a check for TEST_NUMPY

**Comment by cpuhrsch** [comment]
@vedanuj Could you please also verify that the performance for doubles hasn't changed?

**Comment by vedanuj** [comment]
Last commit 418d74f adds separate test with Numpy check. 

@cpuhrsch Performance for doubles improves slightly with this change. 

On master
```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 35.51650547981262      type: torch.DoubleTensor
tanh:           size: 10^3      count: 1000     elapsed: 10.246849060058594     type: torch.DoubleTensor
tanh:           size: 10^4      count: 100      elapsed: 5.211124897003174      type: torch.DoubleTensor
tanh:           size: 10^5      count: 10       elapsed: 7.016204595565796      type: torch.DoubleTensor
```

After 418d74f :
```
['tanh']
tanh:           size: 10^2      count: 10000    elapsed: 25.467238664627075     type: torch.DoubleTensor
tanh:           size: 10^3      count: 1000     elapsed: 8.186361312866211      type: torch.DoubleTensor
tanh:           size: 10^4      count: 100      elapsed: 4.314857721328735      type: torch.DoubleTensor
tanh:           size: 10^5      count: 10       elapsed: 2.8747189044952393     type: torch.DoubleTensor
```

**Comment by vedanuj** [comment]
Opened a new [PR](https://github.com/pytorch/pytorch/pull/7495) using Sleef's `expf`. Closing this.
